### PR TITLE
Block Moving Mode: Prevent the block from being moved into itself

### LINKED ISSUE #54016: Browser freezes when moving into its own block in "Move to" mode
### Description

Select Move to from the block toolbar and a blue indicator will appear. This indicator can be manipulated using the arrow keys to know where to move the block.

When I use this indicator and move into the block itself that I am trying to move, the browser freezes. In the environments I have monitored, the CPU is at 100%, and clicking anywhere on the screen becomes unresponsive.

### Step-by-step reproduction instructions

- Insert a Group block.
- Insert some blocks in the group block.
- Select a group block.
- Select "Move to" from the block toolbar.
- Press the right arrow key. This means move into the group block.
- Press the Enter key.
- Your browser should freeze.

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/54422211/f9f45461-0c66-46ec-8991-91e3cb17f560

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes: #54016

## What?
This PR ensures that when a block is moved in Move to mode, it cannot be moved inside its own block.

## Why?
The block is not supposed to be moved into itself, because if it were, an infinite loop would occur and the browser would freeze.

## How?

When the block itself to be moved is selected, pressing the right key cancels the attempt to enter the block. At the same time, we added basic E2E tests.

### Testing Instructions for Keyboard

- Insert two group blocks with multiple paragraphs.
- Select the first group block.
- Select "Move to" from the block toolbar.
- Press the right key to move it into itself. Nothing should happen.
- Press the down key to move to the next group block.
- Press the right key and it should move correctly into the group block.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/c296bb9c-461d-4181-a2cf-60597e7b310a

### CODE REVIEW

File: packages/block-editor/src/components/block-tools/block-selection-button.js [hunk 1]
```diff
@@ -193,6 +193,13 @@ function BlockSelectionButton( { clientId, rootClientId } ) {
 			selectBlock( startingBlockClientId );
 			setBlockMovingClientId( null );
 		}
+		if (
+			startingBlockClientId &&
+			selectedBlockClientId === startingBlockClientId &&
+			navigateIn
+		) {
+			return;
+		}
```
Comment (Mamaduka) [inline packages/block-editor/src/components/block-tools/block-selection-button.js:203]: Let's add the comment above this check so it's clear why we're returning early.

File: test/e2e/specs/editor/various/block-moving-mode.spec.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		await editor.showBlockToolbar();
+		await page
+			.getByRole( 'toolbar', { name: 'Block tools' } )
+			.getByRole( 'button', { name: 'Options' } )
+			.click();
+		await page.getByRole( 'menuitem', { name: 'Move to' } ).click();
+		await page.keyboard.press( 'ArrowUp' );
+		await page.keyboard.press( 'Enter' );
+
+		expect( await editor.getEditedPostContent() ).toMatchSnapshot();
```
Comment (Mamaduka) [inline test/e2e/specs/editor/various/block-moving-mode.spec.js]: There's a plan to move away from snapshot testing - https://developer.wordpress.org/block-editor/contributors/code/testing-overview/e2e/overusing-snapshots/#the-solution.

**Review by Mamaduka — APPROVED** [review]
Thanks, @t-hamano. The fix works well in my tests ✅

Let's just two small note as a suggestion!

File: test/e2e/specs/editor/various/block-moving-mode.spec.js (hunk 2 above)
Comment (t-hamano) [reply test/e2e/specs/editor/various/block-moving-mode.spec.js]: Done ✅

File: packages/block-editor/src/components/block-tools/block-selection-button.js (hunk 1 above)
Comment (t-hamano) [reply packages/block-editor/src/components/block-tools/block-selection-button.js:203]: Done ✅

**Pull request merged**